pkg/storage/stores/shipper/uploads/table_manager.go

```diff
@@ -0,0 +1,230 @@
+package uploads
+
+import (
+	"context"
+	"errors"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"sync"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/chunk"
+	"github.com/cortexproject/cortex/pkg/chunk/local"
+	chunk_util "github.com/cortexproject/cortex/pkg/chunk/util"
+	pkg_util "github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type Config struct {
+	Uploader       string
+	IndexDir       string
+	UploadInterval time.Duration
+}
+
+type TableManager struct {
+	cfg             Config
+	boltIndexClient BoltDBIndexClient
+	storageClient   StorageClient
+
+	metrics   *metrics
+	tables    map[string]*Table
+	tablesMtx sync.RWMutex
+
+	done chan struct{}
+	wg   sync.WaitGroup
+}
+
+func NewTableManager(cfg Config, boltIndexClient BoltDBIndexClient, storageClient StorageClient, registerer prometheus.Registerer) (*TableManager, error) {
+	tm := TableManager{
+		cfg:             cfg,
+		boltIndexClient: boltIndexClient,
+		storageClient:   storageClient,
+		metrics:         newMetrics(registerer),
+		done:            make(chan struct{}),
+	}
+
+	tables, err := tm.loadTables()
+	if err != nil {
+		return nil, err
+	}
+
+	tm.tables = tables
+	return &tm, nil
+}
+
+func (tm *TableManager) loop() {
+	defer tm.wg.Done()
+
+	syncTicker := time.NewTicker(tm.cfg.UploadInterval)
+	defer syncTicker.Stop()
+
+	for {
+		select {
+		case <-syncTicker.C:
+			err := tm.uploadTables(context.Background())
+			if err != nil {
+				level.Error(pkg_util.Logger).Log("msg", "error uploading local boltdb files to the storage", "err", err)
+			}
+		case <-tm.done:
+			return
+		}
+	}
+}
+
+func (tm *TableManager) Stop() {
+	close(tm.done)
+	tm.wg.Wait()
+
+	err := tm.uploadTables(context.Background())
```

Comment:
Same here. plus both of  `uploadTables` references use it only with Background.

Reply:
Same here, we can have a 10 mins timeout on the context.

Reply:
replied above

Reply:
Can you please check if you missed posting the comment? I can't see it yet.